File: src/controllers/simple.js
export class SimpleController {
    constructor(target_temp) {
        this.target_temp = target_temp;
        this.currently_on = true;
    }

    update({ temperature }) {
        if (this.currently_on && temperature > this.target_temp + 0.1) {
            this.currently_on = false;
        } else if (!this.currently_on && temperature < this.target_temp - 0.1) {
            this.currently_on = true;
        }

        return this.currently_on ? 1.0 : 0.0;
    }
}

File: wasm/src/controllers/simple.rs
use crate::controllers::Controller;

pub struct SimpleController {
    target_temperature: f32,
}

impl Controller for SimpleController {
    fn set_target_temperature(&mut self, temperature: f32) {
        self.target_temperature = temperature;
    }

    fn update(&mut self, temperature: f32) -> f32 {
        if temperature > self.target_temperature {
            0.0
        } else {
            1.0
        }
    }
}

impl SimpleController {
    pub fn new() -> SimpleController {
        SimpleController {
            target_temperature: 0.0,
        }
    }
}
